src/odemis/driver/technolution.py

```diff
@@ -534,60 +543,48 @@ def _setCalibrationMode(self, mode):
-                    "(only the characters '%s' are allowed)." % ASM_PASSWORD_ILLEGAL_CHARS[2:-1])
-
-        if re.search(ASM_HOST_ILLEGAL_CHARS, url_parser.hostname):
-            raise ValueError(
-                    "Host contains invalid characters, host remains unchanged "
-                    "(only the characters '%s' are allowed)." % ASM_HOST_ILLEGAL_CHARS[2:-1])
-
-        if re.search(ASM_PATH_ILLEGAL_CHARS, url_parser.path):
-            raise ValueError("Path on ftp server contains invalid characters, path remains unchanged "
-                             "(only the characters '%s' are allowed)." % ASM_PATH_ILLEGAL_CHARS[2:-1])
+        # check that all sub-elements exist. There are special cases where the parser fails splitting correctly.
+        # This can happen, when for example an extra '@' is used after the first one. Then the parser works
+        # incorrectly and sub-elements are None or empty strings after splitting the url input.
+        if not url_parser.scheme or not url_parser.username or not url_parser.password \
+           or not url_parser.hostname or not url_parser.path:
+            raise ValueError("URL %s scheme is incorrect. Must be of format: "
+                             "'ftp://username:password@127.0.0.1:5000/directory/sub-directory'." % url)
+
+        # check the scheme is correct
+        if url_parser.scheme != 'ftp':
+            raise ValueError("URL %s scheme is incorrect. Must be: 'ftp'." % url_parser.scheme)
+
+        if re.search(ASM_USER_CHARS, url_parser.username):
+            raise ValueError("Username %s contains invalid characters. Only the following characters are allowed: "
+                             " '%s'." % (url_parser.username, ASM_USER_CHARS[2:-1]))
+
+        if re.search(ASM_PASSWORD_CHARS, url_parser.password):
+            raise ValueError("Password %s contains invalid characters. Only the following characters are allowed: "
+                             "'%s'." % (url_parser.password, ASM_PASSWORD_CHARS[2:-1]))
+
+        if re.search(ASM_HOST_CHARS, url_parser.hostname):
+            raise ValueError("Host %s contains invalid characters. Only the following characters are allowed: "
+                             "'%s'." % (url_parser.hostname, ASM_HOST_CHARS[2:-1]))
+
+        if re.search(ASM_PATH_CHARS, url_parser.path):
+            raise ValueError("Path %s contains invalid characters. Only the following characters are allowed: "
+                             "'%s'." % (url_parser.path, ASM_PATH_CHARS[2:-1]))
 
         # TODO: Commented out because not present on EA
         # self.asmApiPostCall("/config/set_external_storage?host=%s&user=%s&password=%s" %
         #                     (urlparse(self.externalStorageURL.value).hostname,
         #                      urlparse(self.externalStorageURL.value).username,
         #                      urlparse(self.externalStorageURL.value).password), 204)
+
```

Comment:
I had never noticed, but it's actually not possible to change any of these values... So let's just mark it so by having the externalStorageURL VA as "read-only". You'd still need to validate the input, for the initial value, but the setter would never need to be called afterwards.

Reply:
In the API we have this call, which is not implemented yet. 
```
  /config/set_external_storage:
    post:
      tags: ["config"]
      summary: "Configure the external storage FTP server. [NOT IMPLEMENTED] Is configured via configuration file on the ASM."
      description: "Configure the external FTP server for storage of the mega field on the server by providing the server address and necessary login credentials."
      ....
```
Do  you mean that we then never ever will make use of it? And that there is no need to implement this at all also in future? Why was it in the design of the API in first place?

Reply:
I think that if the API becomes available, then we can remove the "read-only" flag from the VA (in addition to actually calling the API function). 
Probably the intention is to allow the user choose a different storage sever without having to modify the ASM config (which only a service engineer can do). I'm fine with this not being implemented yet, but it's probably good to keep it mind and not wait too long to ask Technolution to implement it.
For now, if the value cannot be changed, it's better to indicate it, by using the "readonly" flag on the VA.